### PR TITLE
Fix long file paths on Windows

### PR DESCRIPTION
**Pull request by atilaneves**
dub uses relative paths when compiling dependent dub packages, which sometimes result in file paths that are too long for `mkdir` on Windows. This patch fixes the issue by using the newer unicode Win32 path APIs to create directories and files in dmd.

### CODE REVIEW

File: src/ddmd/root/file.d [hunk 1]
```diff
@@ -256,7 +256,12 @@ nothrow:
         {
             DWORD numwritten;
             const(char)* name = this.name.toChars();
-            HANDLE h = CreateFileA(name, GENERIC_WRITE, 0, null, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_SEQUENTIAL_SCAN, null);
+            HANDLE h = CreateFileW(&name.toExtendedLengthPath[0],
+                                   GENERIC_WRITE,
+                                   0,
+                                   null,
+                                   CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_SEQUENTIAL_SCAN,
```
Comment (jmdavis) [inline src/ddmd/root/file.d]: Just a stylistic nitpick, but this would be clearer if `FILE_ATTRIBUTE_NORMAL | FILE_FLAG_SEQUENTIAL_SCAN,` went on its own line. With everything else on its own line, at first glance, it looks like `CREATE_ALWAYS` is being |ed with the other two even though it isn't.

File: src/ddmd/root/filename.d [hunk 2]
```diff
@@ -27,7 +27,9 @@ import ddmd.root.rootobject;
 nothrow
 {
 version (Windows) extern (C) int mkdir(const char*);
-version (Windows) alias _mkdir = mkdir;
+version (Windows) extern(Windows) uint CreateDirectoryW(const(wchar)*, void*);
+version (Windows) extern(Windows) uint GetFullPathNameW(const(wchar)*, uint, const(wchar)*, const(wchar)*) pure;
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: I'm not convinced that this can be `pure`. In particular, the section of the docs that starts with "Multithreaded applications and shared library code should not use the GetFullPathName function..." strongly implies that it depends on state beyond what's passed to it.

https://msdn.microsoft.com/en-us/library/windows/desktop/aa364963(v=vs.85).aspx

File: src/ddmd/root/filename.d [hunk 3]
```diff
@@ -27,7 +27,9 @@ import ddmd.root.rootobject;
...
+version (Windows) extern(Windows) uint GetLastError() @safe pure;
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: How on earth can this be `pure`? It changes pretty much as soon as you call it. In fact, my experience is that unless you call it _immediately_ after the function that should be setting the error failed, it reports that there was no error even though you know that there was.

File: src/ddmd/root/filename.d [hunk 4]
```diff
@@ -721,3 +724,111 @@ nothrow:
...
+    private int _mkdir(const(char)* path) nothrow
+    {
+        import core.sys.windows.winerror: ERROR_ALREADY_EXISTS;
+        import core.stdc.errno: errno, EEXIST;
+
+        const extendedPath = path.toExtendedLengthPath();
+
+        // the CreateDirectoryW function takes a pointer to a structure,
+        // but since we're not interested in using it it's declared void*
+        // and we pass null for the default behaviour.
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: Then why not just literally pass it `null`? Why the long comment and the variable that does nothing?

File: src/ddmd/root/filename.d [hunk 5]
```diff
@@ -721,3 +724,111 @@ nothrow:
...
+        void* securityAttributes;
+        const createRet = () @trusted
+        {
+            return CreateDirectoryW(&extendedPath[0], securityAttributes);
+        }();
+        const lastError = GetLastError();
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: My experience with `GetLastError` has been that if you put _any_ code between it and the function that failed, then it will report that there was no error (e.g. for cross-platform socket code, I was forced to use a string mixin rather than a helper function, or I always got the wrong answer from `GetLastError`). I think that there's a good chance that this will report incorrectly due to the lambda. So, I think that this needs to be handled differently, as annoying as that may be.

File: src/ddmd/root/filename.d [hunk 6]
```diff
@@ -721,3 +724,111 @@ nothrow:
...
+        if (createRet == 0 && lastError == ERROR_ALREADY_EXISTS)
+            errno(EEXIST);
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts an UTF8 null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    private wchar[] toWStringz(const(char*) str) pure nothrow {
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: I'm tempted to argue that you should be porting over `tempCString` from std.internal.cstring, but I think that for that to work, it has to be passed to something within the function that calls it rather than returned from the function. That could probably still be done with the string being constructed at the point that it's passed to `CreateFileW`, but that would probably end up being a lot less straightforward, and `tempCString` seems to use `byUTF`, which isn't available here. So, while this isn't ideal, it's probably fine.

File: src/ddmd/root/filename.d [hunk 7]
```diff
@@ -721,3 +724,111 @@ nothrow:
...
+        // Now wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+
+        static immutable prefix = `\\?\`w; // see comment above
+
+        // We don't know how large the absolute file path will be We
+        // need to tell GetFullPathName the size of the buffer we're
+        // passing in so we defensively allocate 4 times the size of
+        // the string we have currently.
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: I confess that this makes me nervous, but the way to do this that I can think of where you can guarantee correctness is to make it the length of the maximum path length, and per the docs, that's now 32,767, which would be rather excessive. 4x seems like it would be fine for most cases, but I don't know if there are corner cases where it would be a problem. I can't think of any though.

File: src/ddmd/root/filename.d [hunk 8]
```diff
@@ -721,3 +724,111 @@ nothrow:
...
+        {
+            return GetFullPathNameW(&wpath[0],
+                                    absPath.length,
+                                    &absPath[prefix.length],
+                                    null /*filePartBuffer*/);
+        }();
+
+        if (absPathRet > absPath.length) return null;
+
+        return absPath[0 .. absPathRet];
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: Personally, I think that it's cleaner (and it's definitely shorter) to just turn these lines into a single return with a ternary operator, but I don't know if the dmd devs would care one way or the other.

File: src/ddmd/root/filename.d (hunk 6 above)
Comment (jmdavis) [reply src/ddmd/root/filename.d]: Actually, on further reflection, what I'd be tempted to do would be to implement `toExtendedLengthPath` as a function that took a `wchar[]` as a buffer and then pass it a slice of a large static array and try to implement that such that it didn't allocate anything (particularly in light of the fact that dmd is kind of sucky with regards to memory usage), but I'm not sure that that's worth it, and it would max out the path length at whatever the static array's length was, which could be large, but it would still limit it more than would be the case with a dynamic array.

But if the dmd devs have no problem with constructing the dynamic array like this, then this should be fine.

File: src/ddmd/utils.d [hunk 9]
```diff
@@ -92,6 +92,7 @@ extern (C++) void writeFile(Loc loc, File* f)
  */
 extern (C++) void ensurePathToNameExists(Loc loc, const(char)* name)
 {
+    import core.stdc.stdio;
```
Comment (jmdavis) [inline src/ddmd/utils.d]: Why was this added? A deprecation message about imports? Certainly, you didn't add anything here which uses the import.

File: src/ddmd/root/filename.d (hunk 2 above)
Comment (jmdavis) [reply src/ddmd/root/filename.d]: Also, aside from trying to tweak the attributes (which I don't think is actually valid here), couldn't you just import these from `core.sys.windows.winbase`?

File: src/ddmd/root/file.d (hunk 1 above)
Comment (atilaneves) [reply src/ddmd/root/file.d]: I made a mistake while separating.

File: src/ddmd/root/filename.d (hunk 2 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: At the time I didn't know these were in winbase, and then later when I found out I forgot to go back and delete these ad-hoc declarations.

File: src/ddmd/root/filename.d (hunk 3 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Oops.

File: src/ddmd/root/filename.d (hunk 4 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: One could argue that either the comment or the variable should exist but not both. The idea is that a caller seeing `foo(x, y, null);` has no idea what `null` means in that context - it's a magic value. Given the comment here, I'll delete the variable.

File: src/ddmd/root/filename.d (hunk 5 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Fixed.

File: src/ddmd/root/filename.d (hunk 6 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: The problem is, as you stated, picking that size. It completely depends on what happens at runtime. The best solution would probably be to use a static array and allocate if it needs more memory. I've changed it to be a little bit more complicated but faster for usual workloads.

File: src/ddmd/root/filename.d (hunk 8 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Me too, I don't know why I wrote it like this.

File: src/ddmd/utils.d (hunk 9 above)
Comment (atilaneves) [reply src/ddmd/utils.d]: Leftover from printf debugging.

File: src/ddmd/root/filename.d (hunk 4 above)
Comment (jmdavis) [reply src/ddmd/root/filename.d]: I honestly don't see any point in having this comment - especially when you have a comment next to the `null`. This comment really doesn't tell you anything more than should be obvious from the `null` argument, and the comment right next to the `null` gives you some idea what it is (more than this comment does) - though honestly, I think that if anyone wants to really know what the `null` is for needs to read the docs anyway.

File: src/ddmd/root/filename.d [hunk 10]
```diff
@@ -680,6 +679,8 @@ nothrow:
         }
         else version (Windows)
         {
+            import core.sys.windows.winbase: GetFullPathNameA;
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: Why is something here using `GetFullPathNameA` instead of `GetFullPathNameW`? Doesn't that just invite the sort of the problems that this PR is supposed to fix?

File: src/ddmd/root/filename.d [hunk 11]
```diff
@@ -721,3 +722,124 @@ nothrow:
...
+
+        const absPathRet = () @trusted
+        {
+            return GetFullPathNameW(&wpath[0],
+                                    absPath.length,
+                                    &absPath[prefix.length],
+                                    null /*filePartBuffer*/);
+        }();
+
+        return absPathRet > absPath.length ? null : absPath[0 .. absPathRet];
```
Comment (jmdavis) [inline src/ddmd/root/filename.d]: This could be returning a slice of `absBuf`, which is _really_ bad, because it's a local static array. If you want to return a slice of a static array, then you'll need to do something like reuse `buf` somehow.

**Review by jmdavis — CHANGES_REQUESTED** [review]
You've managed to reduce the allocations with the static arrays, which is good, but you managed to return a slice of a static array when the path is short enough, which is not at all `@safe`.

File: src/ddmd/root/filename.d (hunk 4 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Removed.

File: src/ddmd/root/filename.d (hunk 10 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: It does, and I wondered about this myself. The problem is that I'm not familiar with this code or what it does and it's used in a bunch of places.

File: src/ddmd/root/filename.d (hunk 11 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Sigh, I forgot to make it static, good catch. I guess reusing `buf` is a solution.

Comment (atilaneves) [reply src/ddmd/root/filename.d]: Changed it so it doesn't return the slice now.

**Review by Laeeth — COMMENTED** [review]
Hey - do you check as I suggested to see if the path already begins with \\\ ?  Because it's not totally uncommon to specify a path to a remote resource rather than a drive on Windows and currently I guess this might work, and it won't after the fix.

File: src/ddmd/root/filename.d (hunk 10 above)
Comment (Laeeth) [reply src/ddmd/root/filename.d]: Isn't this what you want?
https://docs.microsoft.com/en-gb/cpp/c-runtime-library/reference/fullpath-wfullpath
_wfullpath

Comment (Laeeth) [reply src/ddmd/root/filename.d]: @jmdavis @atilaneves ping

File: src/ddmd/root/filename.d (hunk 7 above)
Comment (Laeeth) [reply src/ddmd/root/filename.d]: If you call _wfullpath with a null pointer as a pointer to the buffer to store the absolute path then it calls malloc for you, and you just need to free the result.  Presuming that the behaviour of _wfullpath is identical in this respect to _fullpath, which I think is true.

https://docs.microsoft.com/en-gb/cpp/c-runtime-library/reference/fullpath-wfullpath
"The _fullpath function expands the relative path name in relPath to its fully qualified or absolute path and stores this name in absPath. If absPath is NULL, malloc is used to allocate a buffer of sufficient length to hold the path name. It is the responsibility of the caller to free this buffer. "

I might have misunderstood what this does - possibly it's not what we want.  (I'm not on a Windows machine right now).  But I'm not a Windows guy and yet it didn't take long to find the function (and as so often the case, when you do it the right way it solves other problems you didn't think were related).  I just figured the function must exist, and googled for it.  I had the advantage of knowing I didn't know Windows API at all, which meant I didn't make assumptions - so if we could move a bit towards operating that way more it would help, I think.

Comment (Laeeth) [reply src/ddmd/root/filename.d]: Operating that way meaning identifying what the problem is precisely, and checking to see if there's a better answer.

File: src/ddmd/root/filename.d (hunk 10 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: @Laeeth No. All I did here was add a local import instead of depending on the module. Also, the Win32 API is preferable to Microsoft's C runtime.

**Comment by atilaneves** [comment]
@Laeeth No, I don't check that yet.

File: src/ddmd/root/filename.d [hunk 12]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+    private wchar[] toWStringz(const(char*) str, wchar[] buf = []) pure nothrow
+    {
+        import core.stdc.string: strlen;
+
+        const length8 = strlen(str);
+
+        // Conservative estimate that should be enough for any UTF8 string
+        // according to https://tools.ietf.org/html/rfc3629.
+        // The +1 is for the null terminator.
+        const length16 = length8 * 4 + 1;
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: I don't get it. Worst case utf-8 to utf-16 is a x2 blowup if all characters are in 0 - 127, right?

File: src/ddmd/root/filename.d [hunk 13]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+        int wstrLen;
+        try
+            foreach(wchar c; str[0 .. length8]) wstr[wstrLen++] = c;
+        catch(Exception)
+            return null;
+
+        wstr[wstrLen] = 0; // null-terminate it
+
+        return wstr[0 .. wstrLen];
+    }
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d:874]: can't use https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072(v=vs.85).aspx ?

File: src/ddmd/root/filename.d (hunk 12 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d]: and that's only in the number of bytes. Given that wchar is double the size of char, the correct line would be `const length16 = length8 + 1`;

File: src/ddmd/root/filename.d [hunk 14]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+        auto extendedPath = absPathRet > absPath.length ? null : absPath[0 .. absPathRet];
+        return F(extendedPath);
+    }
+
+
+    // Converts an UTF8 null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    // buf is passed as a scratch space to store the result. If more memory
+    // is needed then toWstringz allocates on the GC heap instead.
+    private wchar[] toWStringz(const(char*) str, wchar[] buf = []) pure nothrow
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d:849]: good time to use `return` a la DIP25?

File: src/ddmd/root/filename.d [hunk 15]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+        import core.stdc.errno: errno, EEXIST;
+
+        const createRet = path.extendedPathThen!(p => CreateDirectoryW(&p[0],
+                                                                       null /*securityAttributes*/));
+        const lastError = GetLastError();
+
+        // Preserve compatibility with mkdir since the calling code expects
+        // errno to be set.
+        if (createRet == 0 && lastError == ERROR_ALREADY_EXISTS)
+            errno(EEXIST);
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: but not on any other error? `ERROR_PATH_NOT_FOUND` -> `ENOENT`

File: src/ddmd/root/filename.d [hunk 16]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: How does this compare to https://docs.microsoft.com/en-gb/cpp/c-runtime-library/reference/mkdir-wmkdir ?

File: src/ddmd/root/filename.d [hunk 17]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+
+        wchar[1024] wpathBuf;
+        // Since the unicode Win32 APIs use UTF16, we need to convert
+        // from UTF8 to UTF16 first.
+        const wpath = path.toWStringz(wpathBuf);
+
+        // Now wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: `static`?

File: src/ddmd/root/filename.d [hunk 18]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+
+        // We don't know how large the absolute file path will be.
+        // We need to tell GetFullPathNameW the size of the buffer we're
+        // passing in so we defensively allocate 4 times the size of
+        // the string we have currently.
+        const absLength = prefix.length + wpath.length * 4;
+        static wchar[1024] absBuf;
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: `static`?

File: src/ddmd/root/filename.d [hunk 19]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+    */
+    private int _mkdir(const(char)* path) nothrow
+    {
+        import core.sys.windows.winbase: CreateDirectoryW, GetLastError;
+        import core.sys.windows.winerror: ERROR_ALREADY_EXISTS;
+        import core.stdc.errno: errno, EEXIST;
+
+        const createRet = path.extendedPathThen!(p => CreateDirectoryW(&p[0],
+                                                                       null /*securityAttributes*/));
+        const lastError = GetLastError();
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: `CreateDirectoryW` doesn't necessarily set the error code on success (see [GetLastError](https://msdn.microsoft.com/en-us/library/windows/desktop/ms679360(v=vs.85).aspx) Return Value section), so you might be reading an error code from a previous windows call. Have to call `SetLastError(NO_ERROR)` before. Windows :(

File: src/ddmd/root/filename.d [hunk 20]
```diff
@@ -721,3 +722,112 @@ nothrow:
...
+        // Now wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
+
+        // We don't know how large the absolute file path will be.
+        // We need to tell GetFullPathNameW the size of the buffer we're
+        // passing in so we defensively allocate 4 times the size of
+        // the string we have currently.
+        const absLength = prefix.length + wpath.length * 4;
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: why not just ask? See https://github.com/dlang/phobos/pull/4938/files#diff-8ac4ebb5dcc368ac0e9b7b9f118aa824R3174 for a similar situation.

edit: better link https://github.com/dlang/phobos/blob/v2.076.1/std/process.d#L3537

File: src/ddmd/root/filename.d (hunk 12 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: After thinking about it for a while, you're right about the correct line, but maybe not the explanation. I completely confused length in `wchar` with number of bytes, but the way to reason about is this: _if_ the UTF16 encoding needs 2 code units, then the corresponding UTF8 encoding needs at least 2 bytes, so the worst case scenario is UTF16 has a one-to-one correspondence in the string lengths.

File: src/ddmd/root/filename.d (hunk 13 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d:874]: What would I pass as the code page?

File: src/ddmd/root/filename.d (hunk 14 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d:849]: DIP25 only applies to references, no?

File: src/ddmd/root/filename.d (hunk 15 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Fixed.

File: src/ddmd/root/filename.d (hunk 16 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Those two have the 248 character limitation addressed by this PR.

File: src/ddmd/root/filename.d (hunk 17 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: So it's not put on the stack. It's nearly a fancy version of `enum`.

File: src/ddmd/root/filename.d (hunk 18 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: The code used to return a slice of that, now that it doesn't any more I removed the `static`.

File: src/ddmd/root/filename.d (hunk 19 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Sigh... fixed.

File: src/ddmd/root/filename.d (hunk 20 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: I hadn't considered that, great idea.

File: src/ddmd/root/filename.d (hunk 12 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d]: There are no code points in utf16 that require more code units than the same code point in utf8. I think my reasoning before is right as well, the only case where you have to use the same number of code units in utf16 is for the first 128

File: src/ddmd/root/filename.d (hunk 13 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d:874]: `CP_UTF8`? The magical transcoding loop (iterating `char[]` by `wchar`s) you're using a few lines above is assuming that anyway, right?

File: src/ddmd/root/filename.d (hunk 16 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d]: Ah, ok

File: src/ddmd/root/filename.d [hunk 21]
```diff
@@ -680,20 +677,36 @@ nothrow:
...
+                length16 = GetFullPathNameW(&wpath[0], length16, buf, null);
+                if (length16 == 0)
                 {
                     .free(buf);
                     return null;
                 }
-                return buf;
+
+                // allocate enough space for a UTF8 encoding of buf
+                const length8 = length16 * 4 + 1;
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: should be `3` not `4`, worst case for translating a code point from `utf-16` to `utf-8` is 3 `utf-8` code units for 1 `utf-16` code unit, which happens for code points `0x000800 - 0x00FFFF`

File: src/ddmd/root/filename.d [hunk 22]
```diff
@@ -680,20 +677,36 @@ nothrow:
...
+                auto str = cast(char*)malloc(length8);
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: check return value of `malloc`.

Also do the following work with a slice not a raw pointer?

Comment (John-Colvin) [reply src/ddmd/root/filename.d]: Is there a reason to use `malloc` here instead of `new`?

File: src/ddmd/root/filename.d [hunk 23]
```diff
@@ -680,20 +677,36 @@ nothrow:
...
+                size_t strLen;
+
+                try
+                    foreach(char c; buf[0 .. length16]) str[strLen++] = c;
+                catch(Exception _) {}
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: this seems bad. Don't you mean `catch(Exception _) { return null; }`?

File: src/ddmd/root/filename.d (hunk 14 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d:849]: yup, sorry.

File: src/ddmd/root/filename.d (hunk 13 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d:874]: Still wondering about this.

File: src/ddmd/root/filename.d [hunk 24]
```diff
@@ -721,3 +734,142 @@ nothrow:
...
+                case ERROR_PATH_NOT_FOUND:
+                    errno(ENOENT);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d:788]: `return !createRet;`?

File: src/ddmd/root/filename.d [hunk 25]
```diff
@@ -721,3 +734,142 @@ nothrow:
...
+        SetLastError(NO_ERROR);
+        const createRet = path.extendedPathThen!(p => CreateDirectoryW(&p[0],
+                                                                       null /*securityAttributes*/));
+        const lastError = GetLastError();
+
+        // Preserve compatibility with mkdir since the calling code expects
+        // errno to be set.
+        if (createRet == 0)
+        {
+            switch (lastError)
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d:774]: `switch (GetLastError())`?

File: src/ddmd/root/filename.d [hunk 26]
```diff
@@ -721,3 +734,142 @@ nothrow:
...
+        // wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
+
+        // +1 for the null terminator
+        const bufferLength = pathLength + prefix.length + 1;
+
+        wchar[1024] absBuf;
+        auto absPath = bufferLength > absBuf.length ? new wchar[bufferLength] : absBuf[];
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d:826]: using `new` here, `malloc` in other places.

File: src/ddmd/root/filename.d [hunk 27]
```diff
@@ -721,3 +734,142 @@ nothrow:
...
+                                            null /*filePartBuffer*/);
+
+        if (absPathRet == 0 || absPathRet > absPath.length)
+        {
+            return F(emptyPath);
+        }
+
+        auto extendedPath = absPath[0 .. absPathRet];
+        return F(extendedPath);
+    }
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d:842]: There're quite a few variables that feel unnecessary here (`extendedPath`, `emptyPath`, `absPathRet`, `bufferLength`), but I guess it's a matter of taste.

File: src/ddmd/root/filename.d [hunk 28]
```diff
@@ -721,3 +734,142 @@ nothrow:
...
+        auto absPath = bufferLength > absBuf.length ? new wchar[bufferLength] : absBuf[];
+
+        absPath[0 .. prefix.length] = prefix[];
+
+        const absPathRet = GetFullPathNameW(&wpath[0],
+                                            absPath.length - prefix.length,
+                                            &absPath[prefix.length],
+                                            null /*filePartBuffer*/);
+
+        if (absPathRet == 0 || absPathRet > absPath.length)
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: `absPathRet > absPath.length` can't really happen unless there's a bug, but also isn't the correct error check because you didn't pass a buffer of length `absPath.length`, you passed `absPath.length - prefix.length`. Even then, it really shouldn't be able to happen.

**Comment by John-Colvin** [comment]
pls push the latest upstream to master so the diff isn't mental (or give me write access so I can)

File: src/ddmd/root/filename.d [hunk 29]
```diff
@@ -680,20 +677,36 @@ nothrow:
...
              */
-            DWORD result = GetFullPathNameA(name, 0, null, null);
-            if (result)
+            DWORD length16 = GetFullPathNameW(&wpath[0], 0, null, null);
+            if (length16)
             {
-                char* buf = cast(char*)malloc(result);
-                result = GetFullPathNameA(name, result, buf, null);
-                if (result == 0)
+                auto buf = cast(wchar*)malloc(length16);
```
Comment (John-Colvin) [inline src/ddmd/root/filename.d]: check return value of `malloc`.

**Comment by Laeeth** [comment]
Hey should I rebase and merge?

**Comment by atilaneves** [comment]
@Laeeth No, not yet.

File: src/ddmd/root/filename.d (hunk 21 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Fixed.

File: src/ddmd/root/filename.d (hunk 22 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: The original code used `malloc`, so I kept it that way. Changing to `new`.

File: src/ddmd/root/filename.d (hunk 23 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Yep, changed it.

File: src/ddmd/root/filename.d (hunk 24 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d:788]: I guess, but I like making it explicit.

File: src/ddmd/root/filename.d (hunk 25 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d:774]: I would have done, but Jonathan expressed concern about not immediately calling `GetLastError`.

File: src/ddmd/root/filename.d (hunk 27 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d:842]: `emptyPath` avoids a cast.

File: src/ddmd/root/filename.d (hunk 28 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d]: Do you want me to change it to an assert?

File: src/ddmd/root/filename.d (hunk 13 above)
Comment (atilaneves) [reply src/ddmd/root/filename.d:874]: I'm not convinced it'd be any simpler - there would be two calls to that function (the first to find out the needed space to allocate) along with associated calls to `GetLastError`.

**Comment by atilaneves** [comment]
@John-Colvin I rebased.

File: src/ddmd/root/filename.d (hunk 25 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d:774]: Ok, good point.

File: src/ddmd/root/filename.d (hunk 24 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d:788]: Ok

File: src/ddmd/root/filename.d (hunk 27 above)
Comment (John-Colvin) [reply src/ddmd/root/filename.d:842]: `""w` is an empty `wstring` or do you mean that you need `const` not `immutable`?

Comment (atilaneves) [reply src/ddmd/root/filename.d:842]: I got so used to empty arrays needing a cast that I forgot that strings have special syntax.
